edgelet/aziot-edged/src/image_gc.rs

```diff
@@ -0,0 +1,265 @@
+    }
+}
+
+async fn garbage_collector(
+    runtime: &edgelet_docker::DockerModuleRuntime<http_common::Connector>,
+    image_use_data: ImagePruneData,
+    bootstrap_image_id: String,
+) -> Result<(), EdgedError> {
+    log::info!("Image Garbage Collection starting daily run");
+
+    // track images associated with extant containers
+
+    // first get list of containers on the device, running or otherwise
+    let extant_containers = match ModuleRuntime::list_with_details(runtime).await {
+        Ok(modules) => {
+            let mut image_ids: HashSet<String> = HashSet::new();
+            image_ids.insert(bootstrap_image_id); // bootstrap edge agent image should never be deleted
+
+            for module in modules {
+                // this is effectively the result of calling GET on /containers/json
+                // image_hash() is created from ImageId (ImageId is filled in by GET /containers/json)
+                let id = edgelet_core::Module::config(&module.0)
+                    .image_hash()
+                    .ok_or_else(|| {
+                        EdgedError::from_err(
+                            "error getting image id for running container",
+                            edgelet_docker::Error::GetImageHash(),
+                        )
+                    })?;
+                image_ids.insert(id.to_string());
+            }
+            image_ids
+        }
+        Err(err) => {
+            return Err(EdgedError::new(format!(
+                "Error in image auto-pruning task. Cannot get running modules. Skipping image auto pruning. {}",
+                err
+            )));
+        }
+    };
+
+    let image_map = image_use_data
+        .prune_images_from_file(extant_containers)
+        .map_err(|e| {
+            EdgedError::from_err(
+                "Module image garbage collection failed to prune images from file.",
+                e,
+            )
+        })?;
+
+    // delete images
+    for key in image_map.keys() {
+        if let Err(e) = ModuleRegistry::remove(runtime, key).await {
+            log::error!("Could not delete image {} : {}", key, e);
+        }
+    }
+
+    Ok(())
+}
+
+/* ================================================ HELPER METHODS ================================================ */
+
+fn validate_settings(settings: &ImagePruneSettings) -> Result<(), EdgedError> {
+    if settings.cleanup_recurrence() < Duration::from_secs(60 * 60 * 24) {
+        log::error!(
+            "invalid settings provided in config: cleanup recurrence cannot be less than 1 day."
+        );
+        return Err(EdgedError::from_err(
+            "invalid settings provided in config",
+            edgelet_docker::Error::InvalidSettings(
+                "cleanup recurrence cannot be less than 1 day".to_string(),
+            ),
+        ));
+    }
+
+    let times = settings.cleanup_time();
+    if times.len() != 5 || !times.contains(':') || times.chars().nth(2) != Some(':') {
```

Comment:
can't `chrono`'s `Time` just parse these for you?

Reply:
Sadly, no. I looked. 

If there's an easier/more maintainable way, I'm all for it. I didn't find one, and at some point it was just faster to write the code.

Reply:
```rust
chrono::NaiveTime::parse_from_str(settings.cleanup_time(), "%H:%M")
```
https://docs.rs/chrono/0.4.21/chrono/naive/struct.NaiveTime.html#method.parse_from_str